### ISSUE TITLE
Consider adding a single command to completely synchronize everything with the Pipfile

### ISSUE THREAD

**Issue by ben-g-** [pypa/pipenv#2723]
Say I manually edit my Pipfile -- add some dependencies, remove some dependencies. I would expect that there would be a single command for updating both the Pipfile.lock and the virtual environment to reflect the current state of the Pipfile. As far as I can tell, there isn't.

----
Here's what there is instead (along with some critiques):

**lock**: updates the Pipfile.lock to reflect the current state of the Pipfile

**sync**: installs all packages specified in the Pipfile.lock, but does not uninstall packages *not* specified in Pipfile.lock (which makes the name misleading, since it doesn't entirely synchronize the environment with the Pipfile.lock)

**update**: runs **lock**, then **sync**, i.e. it does not uninstall packages *not* specified in Pipfile, even though it removes them from the Pipfile.lock

**clean**: runs **lock**, then uninstalls packages not specified in Pipfile.lock. (The help text only says that it does the latter, so the fact that it runs **lock** first comes as a surprise. In other words, based on the help text, one thinks it's analogous to **sync**, but it turns out to be analogous to **update**, and there is no analog to **sync**.) It doesn't install packages, even though it adds them to the Pipfile.lock.

In summary: The **sync** command doesn't entirely synchronize anything, which means it's misnamed. And it's weird to have a command that modifies both the Pipfile.lock and the virtual environment yet leaves them unsynchronized -- which is the case for both **update** and **clean**.

----
Given the existing commands, I think that you can completely synchronize everything with the Pipfile by running the following sequence of two commands: **clean**, then **sync**.

So this is mainly a request to add a single command that does what that sequence of two commands does.

But also, secondarily, a suggestion to clean up the existing commands and their documentation.

**Comment by techalchemy** [comment]
`pipenv install` does exactly what you are describing

**Comment by ben-g-** [comment]
@techalchemy: I tested your claim about `pipenv install`, and it is incorrect. It seems to do the same as `pipenv update` in this scenario, i.e.:
- synchronize Pipfile.lock with Pipfile
- then install all packages specified in Pipfile.lock
- but not uninstall packages _not_ specified in Pipfile.lock (e.g. ones it removed from Pipfile.lock in the first step).

I can give you the details of my test if you're interested.

**Comment by techalchemy** [comment]
Nope, not necessary, we don't provide a single command to both add and remove packages from the environment besides `pipenv install` with the `--python` flag (which always creates a new environment)

The idea here as I mentioned is that the environment is disposable, so we are not going through the overhead of trying to maintain its 'cleanliness' at this point.  I can see arguments in favor of doing that, but we aren't prepared to make an API change for this right now (I would reconsider it after we work out the remaining operational challenges around our semantics, which hopefully we should have in the next release or two)

**Comment by rsyring** [comment]
> The idea here as I mentioned is that the environment is disposable, so we are not going through the overhead of trying to maintain its 'cleanliness' at this point.

For what it's worth, this is surprising to me.  I would not have thought pipenv considers environments disposable.  IMO, the environments created on a developers machine are not disposable.  pipenv reuses one if it's available every time there is a shell or run command.  A user would need to manually delete the current environment to get a new one.  In fact, I think the existence of the `run` command shows that pipenv is assuming environments will be long lived.  That is, you set it up with the packages you want and it will be available for later use.

> I can see arguments in favor of doing that

Hopefully to strengthen those arguments, here are a few situations where I think a "sync" command that did what this issue is asking for would be helpful:

* As a developer, when coming back to a project, to make sure my existing environment is in-sync with the current Pipefile and lock file.
* As a developer, for tox environments, when I want the updating of the environment to be as fast as possible.  I'm assuming a true "sync" would be faster than creating a new environment because most packages would be the same and there would be less disk IO.
* When deploying and wanting to update the environment on the server.  Same rational as just above, faster because of less disk IO.

**Comment by uranusjr** [comment]
I am not completely against the idea of sync + clean (maybe something like `sync --clean` would work), but virtual environments _should_ be considered disposable, especially with a reproducing mechanism (locked requirements.txt, or Pipfile.lock). You can choose to not do it this way, of course, but that would be against best practice, and you’re on your own if toolchain development doesn’t go your way.
